cpp/src/arrow/util/decimal.cc

```diff
@@ -831,7 +831,7 @@ static bool RescaleWouldCauseDataLoss(const Decimal128& value, int32_t delta_sca
   }
 
   *result = value * multiplier;
-  return *result < value;
+  return (value < 0) ? *result < value : *result > value;
```

Comment:
I don't think this is the correct check. It should be sufficient to check that the absolute value of `*result` is less than the absolute value of `value`.

Reply:
the check was in fact backward (for entirely stupid reasons). Do you want me to use an actual absolute value method or is it fine now?